src/models/geograph.py

```diff
@@ -443,27 +450,143 @@ def merge_nodes(
-            geometry=polygon,
             class_label=class_label,
+            bounds=polygon.bounds,
+        )
+        data = {"class_label": class_label, "geometry": polygon}
+        missing_cols = {key: None for key in set(self.df.columns) - set(data.keys())}
+        self.df.loc[final_index] = gpd.GeoSeries(
+            {**data, **missing_cols}, crs=self.df.crs
         )
+        self.df = self.df.sort_index()
+        # Add edges from final node to all neighbours of the original
         self.graph.add_edges_from(
-            zip_longest([final_index], adjacency_list, fillvalue=final_index)
+            zip_longest([final_index], adjacency_set, fillvalue=final_index)
+        )
+
+    def add_habitat(
+        self,
+        name: str,
+        valid_classes: set,
+        max_travel_distance: float = 0.0,
+    ) -> None:
+        """
+        Create habitat graph and store it in habitats dictionary.
+
+        Creates a habitat subgraph of the main GeoGraph that only contains edges
+        between nodes in `valid_classes` as long as they are less than
+        `max_travel_distance` apart. All nodes which are not in `valid_classes`
+        are not in the resulting habitat graph.
+
+        Args:
+            name (str): The name of the habitat.
+            valid_classes (set): A set of class labels which make up the habitat.
+            max_travel_distance (float): The maximum distance the animal(s) in
+            the habitat can travel through non-habitat areas. The habitat graph
+            will contain edges between any two nodes that have a class label in
+            `valid_classes`, as long as they are less than `max_travel_distance`
+            units apart. Defaults to 0, which will only create edges between
+            directly neighbouring areas.
+        """
+        hgraph: nx.Graph = deepcopy(self.graph)
+        hgraph.graph["name"] = name
+        # Remove all edges in the graph, then at the end we only have edges
+        # between nodes less than `max_travel_distance` apart
+        hgraph.clear_edges()
+        # Vectorised buffer on the entire df to calculate the expanded polygons
+        # used to get intersections.
+        self.df["buff_polygon"] = self.df.geometry.buffer(max_travel_distance)
+        # List of nodes that are not in the habitat
+        invalid_indexes: List[int] = []
+        for node in tqdm(
+            hgraph.nodes, desc="Generating habitat graph", total=len(hgraph)
+        ):
+            if hgraph.nodes[node]["class_label"] not in valid_classes:
+                invalid_indexes.append(node)
+                continue
+            # df.at is a fast df.loc designed only to access single values
+            polygon = self.df.at[node, "geometry"]
+            buff_poly = self.df.at[node, "buff_polygon"]
+            # Query rtree for all polygons within `max_travel_distance` of the original
+            poss_neighbours = list(self.rtree.intersection(buff_poly.bounds))
```

Comment:
You probably checked this, but this obviously might have quite a few non-relevant neighbours. Have you considered creating an rtree for the habitat? (Just wondering, don't mind it either way - I am not familiar with the complexity of creating an rtree)

Reply:
The fastest way to creating an rtree for the habitat would be to deepcopy the main one and remove all non-habitat nodes. This would take some upfront computation time, so for this to be faster eventually would require both quite a small habitat and for us to frequently use some other function for habitats which iterates through the results of rtree queries. I'm not sure if there will be another function like that, so I think we'll leave it for now.